### PR TITLE
KAFKA-14589 [1/3] Tests of ConsoleGroupCommand rewritten in java

### PR DESCRIPTION
**Pull request by nizhikov**
This PR is part of https://github.com/apache/kafka/pull/14471
Is contains some of `ConsoleGroupCommand` tests rewritten in java.
Intention of separate PR is to reduce changes and simplify review.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by showuon** [comment]
Will take a look next week. Thanks.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+        ConsumerGroupCommand.ConsumerGroupService service = getConsumerGroupService(cgcArgs);
+
+        TestUtils.waitForCondition(
+            () -> service.collectGroupMembers(GROUP, false)._2.get().size() == 1,
+            "The group did not initialize as expected."
+        );
+
+        String output = kafka.utils.TestUtils.grabConsoleOutput(() -> {
+            service.deleteGroups();
+            return null;
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java:103]: Will it have compiling failure after removing `return null`?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+
+        TestUtils.waitForCondition(
+            () -> Objects.equals(service.collectGroupState(GROUP).state(), "Empty"),
+            "The group did not become empty as expected.");
+
+        String[] cgcArgs2 = new String[cgcArgs.length + 2];
+
+        System.arraycopy(cgcArgs, 0, cgcArgs2, 0, cgcArgs.length);
+        cgcArgs2[cgcArgs2.length - 2] = "--group";
+        cgcArgs2[cgcArgs2.length - 1] = missingGroup;
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java]: I know we did that in scala like this, but after rewriting in Java it becomes more unreadable. Maybe we can just do this:
```
String[] cgcArgs2 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--delete", "--group", GROUP, "--group", missingGroup};
```

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java]: ditto

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+    String[] getArgs(String group, String topic) {
+        return new String[] {
+            "--bootstrap-server", bootstrapServers(listenerName()),
+            "--delete-offsets",
+            "--group", group,
+            "--topic", topic
+        };
+    }
+
+    @Test
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Should we add these?
```
    @ParameterizedTest(name = TEST_WITH_PARAMETERIZED_QUORUM_NAME)
    @ValueSource(strings = {"zk", "kraft"})
```

Same for below tests.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        addConsumerGroupExecutor(1);
+
+        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--state"};
+        ConsumerGroupCommand.ConsumerGroupService service = getConsumerGroupService(cgcArgs);
+
+        scala.collection.Set<ConsumerGroupListing> expectedListing = set(Arrays.asList(
+            new ConsumerGroupListing(simpleGroup, true, Optional.of(ConsumerGroupState.EMPTY)),
+            new ConsumerGroupListing(GROUP, false, Optional.of(ConsumerGroupState.STABLE))));
+
+        final scala.collection.Set[] foundListing = new scala.collection.Set[1];
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Could we use java collection here?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+        scala.collection.Set<ConsumerGroupListing> expectedListingStable = set(Collections.singleton(
+            new ConsumerGroupListing(GROUP, false, Optional.of(ConsumerGroupState.STABLE))));
+
+        foundListing[0] = null;
+
+        TestUtils.waitForCondition(() -> {
+            foundListing[0] = service.listConsumerGroupsWithState(set(Collections.singleton(ConsumerGroupState.STABLE))).toSet();
+            return Objects.equals(expectedListingStable, foundListing[0]);
+        }, "Expected to show groups " + expectedListingStable + ", but found " + foundListing[0]);
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: This is hard to read. `foundListing[0] = null` doesn't make sense to me. Please improve it. Thanks.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+        assertEquals(set(Arrays.asList(ConsumerGroupState.DEAD, ConsumerGroupState.COMPLETING_REBALANCE)), result);
+
+        result = ConsumerGroupCommand.consumerGroupStatesFromString("stable");
+        assertEquals(set(Arrays.asList(ConsumerGroupState.STABLE)), result);
+
+        result = ConsumerGroupCommand.consumerGroupStatesFromString("stable, assigning");
+        assertEquals(set(Arrays.asList(ConsumerGroupState.STABLE, ConsumerGroupState.ASSIGNING)), result);
+
+        result = ConsumerGroupCommand.consumerGroupStatesFromString("dead,reconciling,");
+        assertEquals(set(Arrays.asList(ConsumerGroupState.DEAD, ConsumerGroupState.RECONCILING)), result);
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java:111]: I didn't see these tests in scala. Did you add them?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("bad, wrong"));
+
+        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("  bad, Stable"));
+
+        assertThrows(IllegalArgumentException.class, () -> ConsumerGroupCommand.consumerGroupStatesFromString("   ,   ,"));
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java:117]: It seems 1 test is missing.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"zk", "kraft"})
+    public void testListGroupCommand(String quorum) throws Exception {
+        String simpleGroup = "simple-group";
+        addSimpleGroupExecutor(simpleGroup);
+        addConsumerGroupExecutor(1);
+        final String[] out = {""};
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Why can't we declare as `String` directly? 
`final String out = ""`?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+        }, "Expected to find " + GROUP + " in state Stable and the header, but found " + out[0]);
+
+        String[] cgcArgs4 = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list", "--state", "stable"};
+        TestUtils.waitForCondition(() -> {
+            out[0] = kafka.utils.TestUtils.grabConsoleOutput(() -> {
+                ConsumerGroupCommand.main(cgcArgs4);
+                return null;
+            });
+            return out[0].contains("STATE") && out[0].contains(GROUP) && out[0].contains("Stable");
+        }, "Expected to find " + GROUP + " in state Stable and the header, but found " + out[0]);
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Duplicate?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+    @ValueSource(strings = {"zk", "kraft"})
+    public void testListConsumerGroups(String quorum) throws Exception {
+        String simpleGroup = "simple-group";
+        addSimpleGroupExecutor(simpleGroup);
+        addConsumerGroupExecutor(1);
+
+        String[] cgcArgs = new String[]{"--bootstrap-server", bootstrapServers(listenerName()), "--list"};
+        ConsumerGroupCommand.ConsumerGroupService service = getConsumerGroupService(cgcArgs);
+        scala.collection.Set<String> expectedGroups = set(Arrays.asList(GROUP, simpleGroup));
+        final scala.collection.Set[] foundGroups = new scala.collection.Set[1];
```
Comment (showuon) [inline tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Could we use java set? It's weird we compare the 1st element only below. That is, you create a Set, and then, you just verify the 1st element of the Set. Then, why do you need a set if only 1 element?

**Review by showuon — COMMENTED** [review]
Thanks for the PR. Left some comments.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java (hunk 1 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java:103]: Yes.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java (hunk 4 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteOffsetsConsumerGroupCommandIntegrationTest.java]: Great catch. Thanks. Fixed.
This was added 2 weeks ago and I missed to reflect that change in java test version.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 9 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: `out` value changed in java lambda 

`out = grabConsoleOutput()` and only final variable (which can't be changed) can be used inside java lambda.

I can rework String array to usage of `AtomicReference`. 
What do you think?

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java (hunk 2 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java]: Done.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java (hunk 3 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/DeleteConsumerGroupsTest.java]: done

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 5 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: It seems that usage of scala collections is more convinient for now.
But, of course, when java version of command will be merged we will use java collections here and everywhere in test. You can see how this don for already mereged tests:

https://github.com/apache/kafka/pull/14471/files#diff-4e22e01d567ea1b5087aa60354fa127bb25b892b430a6b86ab39a10edd393c0a

Снимок экрана 2024-01-30 в 16 32 26

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 6 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Reworked with `AtomicReference`.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 7 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java:111]: It is fresh changes from 82920ffad00285c11453e64e220ac0ec2666ac54

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 8 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java:117]: Double checked and can't spot missing test. Can you, please, clarify what test are missing?

**Comment by nizhikov** [comment]
@showuon Thanks for review. Answered to all your comments. Please, take a look one more time.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 9 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Reworked to `AtomicReference`.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 10 above)
Comment (nizhikov) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: No, case of parameter differs - "Stable", "stable".

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 8 above)
Comment (showuon) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java:117]: Ah, there's some update in the test. Please ignore it. Thanks.

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 9 above)
Comment (showuon) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Looks good!

File: tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java (hunk 10 above)
Comment (showuon) [reply tools/src/test/java/org/apache/kafka/tools/consumer/group/ListConsumerGroupTest.java]: Ah, that's why.

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
Re-running the CI build.

**Comment by nizhikov** [comment]
CI results looks good for me.

**Comment by nizhikov** [comment]
Hello @showuon 

CI is OK for me. Are you ready to merge this PR or have other questions, comments?

**Comment by showuon** [comment]
@nizhikov , thanks for the reminder, I forgot about this PR.
I just checked https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15256/8 and https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15256/7, the test: `kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()` failed on each job once. But it didn't appear in earlier run, https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15256/4/, as well as the Kafka trunk build: https://ci-builds.apache.org/job/Kafka/job/kafka/job/trunk/2607/ . Could you help check that? Thanks.

**Comment by dajac** [comment]
@nizhikov I merged https://github.com/apache/kafka/pull/15253. I am not sure if it has any impact on this PR. Does it?

**Comment by nizhikov** [comment]
@dajac It has impact, for sure :)

I reflect your changes in the main PR #14471 and will merge them to this PR in the nearest time.

**Comment by nizhikov** [comment]
@showuon I run `DynamicBrokerReconfigurationTest` locally and it passed.
Not sure, how those failures can relate to this PR as I change only specific tests classes.

**Comment by showuon** [comment]
@nizhikov , could you rebase with the latest trunk branch and let's see if it fixed the test? 

> @showuon I run `DynamicBrokerReconfigurationTest` locally and it passed. Not sure, how those failures can relate to this PR as I change only specific tests classes.

OK, let's wait and see if the latest CI build fixes the test.

**Comment by nizhikov** [comment]
> could you rebase with the latest trunk branch and let's see if it fixed the test?

PR currently merged with the latest trunk changes. https://github.com/apache/kafka/pull/15256/commits/74980be0279f908835ef2ce9f5b1c0aabdf8dcec

**Comment by nizhikov** [comment]
@showuon CI results are ready. 

AFAICS other tests of `DynamicBrokerReconfigurationTest` failed in the latest trunk:
https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2609/tests

**Comment by nizhikov** [comment]
Other trunk test run contains `DynamicBrokerReconfigurationTest ` fails, also (but only for 1 environment).

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2610/tests

**Comment by nizhikov** [comment]
@showuon Please, write me, If those failures should be further investigated.

**Comment by nizhikov** [comment]
Hello @showuon 

What do you think? `DynamicBrokerReconfigurationTest` failures related to this PR? Or it can be merged as is?

**Comment by showuon** [comment]
I think the `kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()` is caused by this PR. We need to investigate it. From the error, it said there is a thread leaked:
```
org.opentest4j.AssertionFailedError: Invalid threads: expected 6, got 7: List(data-plane-kafka-socket-acceptor-ListenerName(EXTERNAL)-SASL_SSL-0, data-plane-kafka-socket-acceptor-ListenerName(INTERNAL)-SSL-0, data-plane-kafka-socket-acceptor-ListenerName(CONTROLLER)-PLAINTEXT-0, data-plane-kafka-socket-acceptor-ListenerName(INTERNAL)-SSL-0, data-plane-kafka-socket-acceptor-ListenerName(INTERNAL)-SSL-0, data-plane-kafka-socket-acceptor-ListenerName(EXTERNAL)-SASL_SSL-0, data-plane-kafka-socket-acceptor-ListenerName(EXTERNAL)-SASL_SSL-0) ==> expected: <true> but was: <false>
```

You can try to run your changed test suites 1 by 1, and check that after all test suite completes, is there any network thread leaked. Let me know if you need help.

**Comment by nizhikov** [comment]
@showuon Looks like `kafka.server.DynamicBrokerReconfigurationTest.testThreadPoolResize()` fails in the latest trunk, also:

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2618/tests

So, maybe the fail not caused by this PR?

**Comment by nizhikov** [comment]
@showuon Other trunk run with `kafka.server.DynamicBrokerReconfigurationTest` fails - https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka/detail/trunk/2617/tests

What do you think? Can this PR be merged as is?

**Comment by showuon** [comment]
I just re-run CI this morning, and this time, it [failed 2 tests in all 4 jobs](https://ci-builds.apache.org/job/Kafka/job/kafka-pr/job/PR-15256/11/#showFailuresLink). Trying to run it again now. I can't merge anything introducing test failures, I think you also understand it. It would be helpful if you could help identify the reason why it failed.
```
Build / JDK 8 and Scala 2.12 / kafka.api.SslAdminIntegrationTest.testAsynchronousAuthorizerAclUpdatesDontBlockRequestThreads()
Build / JDK 8 and Scala 2.12 / kafka.api.SslAdminIntegrationTest.testSynchronousAuthorizerAclUpdatesBlockRequestThreads()
```

**Comment by nizhikov** [comment]
Hello @showuon 

I found that `DeleteTopicTest.scala` don't close resources allocated in case of kraft run. 
It was hided by removed tests and lead to the tests failures we observed earlier.

I fixed `DeleteTopicTest` and now it seems that CI is OK.

Take a look one more time, please.

**Comment by nizhikov** [comment]
Hello, @mimaison @jolshan 

Can you, please, take a look.
Looks like this PR are ready to merge.

**Comment by showuon** [comment]
> Hello @showuon
> 
> I found that `DeleteTopicTest.scala` don't close resources allocated in case of kraft run. It was hided by removed tests and lead to the tests failures we observed earlier.
> 
> I fixed `DeleteTopicTest` and now it seems that CI is OK.
> 
> Take a look one more time, please.

Nice find! Thanks for the fix!

**Comment by nizhikov** [comment]
@showuon Thank you for review and merging!

**Pull request merged**